server/backends/chunkstore/chunkstore.go

```diff
@@ -0,0 +1,413 @@
+	return r.chunkIndex == r.startIndex
+}
+
+func (r *chunkstoreReader) copyAndAdvanceOffset(dst, src []byte) int {
+	bytesRead := copy(dst, src)
+	r.advanceOffset(int64(bytesRead))
+	return bytesRead
+}
+
+func (r *chunkstoreReader) copyToReadBuffer(p []byte) int {
+	if r.chunkOff >= len(r.chunk) {
+		return 0
+	}
+	if r.reverse {
+		remainingBytesInChunk := len(r.chunk) - r.chunkOff
+		if len(p) <= remainingBytesInChunk {
+			return r.copyAndAdvanceOffset(p, r.chunk[(remainingBytesInChunk-len(p)):])
+		} else {
+			return r.copyAndAdvanceOffset(p[(len(p)-remainingBytesInChunk):], r.chunk[:remainingBytesInChunk])
+		}
+	} else {
+		return r.copyAndAdvanceOffset(p, r.chunk[r.chunkOff:])
+	}
+}
+
+func (r *chunkstoreReader) nextChunkExists() (bool, error) {
+	return r.chunkstore.chunkExists(r.ctx, r.blobName, r.nextChunkIndex())
+}
+
+func (r *chunkstoreReader) getNextChunk() error {
+	if exists, err := r.nextChunkExists(); err != nil {
+		return err
+	} else if !exists {
+		return fmt.Errorf("Opening %v: %w", chunkName(r.blobName, r.nextChunkIndex()), os.ErrNotExist)
+	}
+	r.chunkIndex = r.nextChunkIndex()
+	r.advanceOffset(int64(len(r.chunk) - r.chunkOff))
+	r.chunkOff -= len(r.chunk)
+	var err error
+	r.chunk, err = r.chunkstore.readChunk(r.ctx, r.blobName, r.chunkIndex)
+	return err
+}
+
+func (r *chunkstoreReader) eof() (bool, error) {
+	if r.chunkOff < len(r.chunk) {
+		return false, nil
+	}
+	exists, err := r.nextChunkExists()
+	return !exists, err
+}
+
+func (r *chunkstoreReader) shiftToFront(p []byte, bytesRead *int) {
+	if *bytesRead < len(p) {
+		copy(p, p[(len(p)-*bytesRead):])
+		for i := *bytesRead; i < len(p); i++ {
+			p[i] = 0
+		}
+	}
+}
+
+func (r *chunkstoreReader) Read(p []byte) (int, error) {
+	bytesRead := r.copyToReadBuffer(p)
+	if r.reverse {
+		defer r.shiftToFront(p, &bytesRead)
+	}
+	for bytesRead < len(p) {
+		err := r.getNextChunk()
+		if r.reverse {
+			bytesRead += r.copyToReadBuffer(p[:(len(p) - bytesRead)])
+		} else {
+			bytesRead += r.copyToReadBuffer(p[bytesRead:])
+		}
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) && !r.noChunksRead() {
+				return bytesRead, io.EOF
+			}
+			return bytesRead, err
+		}
+	}
+	if eof, err := r.eof(); err != nil {
+		return bytesRead, err
+	} else if eof {
+		return bytesRead, io.EOF
+	}
+	return bytesRead, nil
+}
+
+func (r *chunkstoreReader) Seek(offset int64, whence int) (int64, error) {
+	if offset == r.off {
+		return r.off, nil
+	}
+	switch whence {
+	case io.SeekStart:
+	case io.SeekCurrent:
+		offset = r.off + offset
+	case io.SeekEnd:
+		for {
+			if exists, err := r.nextChunkExists(); err != nil {
+				return r.off, err
+			} else if !exists {
+				break
+			}
+			r.getNextChunk()
+		}
+		offset += r.off + int64(len(r.chunk)-r.chunkOff)
+	default:
+		return r.off, syscall.EINVAL
+	}
+
+	if offset < 0 {
+		return r.off, syscall.EINVAL
+	}
+
+	if offset >= r.off-int64(r.chunkOff) && offset < r.off+int64(len(r.chunk)-r.chunkOff) {
```

Comment:
A lot of this chunk /offset math is hard to follow -- I don't think it's wrong necessarily but I also don't have the brainpower to really read this code and check the correctness either :P One way to make it easier to follow would be to add comments any time you modify or check offsets like this?

Reply:
Good point; it's easy to get used to it and forget that this stuff is super obtuse.